### PR TITLE
Try out parallel Travis builds

### PR DESCRIPTION
**Pull request by citizen428**
## What type of PR is this? (check all applicable)

- [X] Optimization

## Description

This PR tries to improve the time our spec runs take on Travis CI by utilizing two VMs, one for system specs and one for everything except system specs.

## Related Tickets & Documents

n/a

## Added tests?

- [X] no, because they aren't needed

## Added to documentation?

- [X] no documentation needed

### CODE REVIEW

**Comment by rhymes** [comment]
Nice! I wonder if there's a way for this:

Screenshot_2020-03-16 Try out parallel Travis builds by citizen428 · Pull Request #6645 · thepracticaldev dev to

to display which of the two tags is the build related to

**Comment by citizen428** [comment]
Not sure about the GH integration, but on Travis it looks like this:

Screen Shot 2020-03-16 at 18 01 30

**Comment by citizen428** [comment]
@benhalpern Yup, "Ran for" is the time the slower build takes to finish, "Total time" is the sum of the two jobs.

File: .travis.yml [hunk 1]
```diff
@@ -45,7 +48,7 @@ before_install:
   - sudo chown -R elasticsearch:elasticsearch /etc/default/elasticsearch
   - sudo systemctl start elasticsearch
 install:
-  - bundle install --path vendor/bundle
+  - bundle install --jobs=3 --retry=3 --path vendor/bundle
```
Comment (mstruve) [inline .travis.yml]: We are running 2 VMs so why does `jobs` here equal 3? I would have expected 2.

+1 for retry!!!

Comment (citizen428) [reply .travis.yml]: This runs inside each VM and configures how many parallel downloads/installs Bundler performs:

```
--jobs=[<number>], -j[<number>]
  The maximum number of parallel download and install jobs. The default is 1.
```

The 3 came straight from the Travis Ruby docks, I didn't think about it too much tbh. Our Bundle is cached anyway, so hopefully, we don't need too many downloads at any given time anyway.

Comment (mstruve) [reply .travis.yml]: OH, thank you for the explanation! I thought it was one for each VM 🤦‍♀  whoops!

**Comment by citizen428** [comment]
Current status: the spec fails only when running exactly like on Travis, i.e. with `bundle exec rspec --tag ~type:system`. Running it in isolation (single spec or whole file) always works. This is not great, as the spec run takes about 18 minutes on my machine. 😢

**Comment by rhymes** [comment]
Closing and reopening to trigger Codeclimate again. Let's see if this 45.6% decrease in coverage is a fluke:

Screenshot_2020-03-18 Try out parallel Travis builds by citizen428 · Pull Request #6645 · thepracticaldev dev to

**Comment by citizen428** [comment]
@rhymes I just assumed that it only takes the output of one job and assumed that I still need to look into using/configuring https://github.com/codeclimate/test-reporter.

**Comment by rhymes** [comment]
@citizen428 ah ok, thanks for the info.

@maestromac found this that might be of help https://docs.codeclimate.com/docs/configuring-test-coverage#section-parallel-tests

File: spec/workers/users/resave_articles_worker_spec.rb [hunk 2]
```diff
@@ -10,9 +10,12 @@
       it "resave articles" do
         user = create(:user)
         article = create(:article, user: user)
+
         old_updated_at = article.updated_at
 
-        worker.perform(user.id)
+        Timecop.travel(1.minute.from_now) do
```
Comment (citizen428) [inline spec/workers/users/resave_articles_worker_spec.rb]: @rhymes Thank you, nice fix!

**Comment by citizen428** [comment]
@maestromac You can add `CC_TEST_REPORTER_ID` via the Travis [settings page](https://travis-ci.com/github/thepracticaldev/dev.to/settings), I think I'd prefer keeping it out of the `.travis.yml` file.

I already prepared the config according to th Travis doc, so the builds will probably fail until we add that variable.

**Comment by maestromac** [comment]
@citizen428 I can't do that because fork repository doesn't have access to env var in settings page. 

https://docs.travis-ci.com/user/environment-variables/#defining-variables-in-repository-settings
>  we do not provide these values to untrusted builds, triggered by pull requests from another repository.

**Comment by rhymes** [comment]
@citizen428 @maestromac I wonder if alternatives to CodeClimate like Coveralls and Codecov can be of help here as we have lots of frictions with codeclimate lately. Maybe worth looking into in the next cooldown period :D

**Comment by benhalpern** [comment]
I feel like we have a good culture of ensuring adequate tests for new PRs such that we might not need the automated check.

All else equal it's nice to have, but if we want to get rid of it in favor of getting faster tests right away I'm cool to get rid of it while we seek an alternative approach.

**Comment by citizen428** [comment]
We can also try adding the key to our `.travis.yml`, I don't think it's overly sensitive. @maestromac You can either push to this branch directly or send me the key.

**Comment by citizen428** [comment]
I think I slowly am getting close to fixing the coverage part, BUT, new problem: if I move `yarn build-storybook` to the end of `after_script` the build gets marked as failed, because `Artifacts support disabled for pull requests` which leads to a non-0 exit status I guess. If I leave it at script, I think it will interfere with the CC coverage upload, as `$TRAVIS_TEST_RESULT` is only 0 if all steps in script succeed. 

This makes me wonder a bit if Travis is really the right place for building and updating the storybook, and if its, if we should run it as a deploy, not during the build phases? 

In the meantime though I think I'm gonna move it back to script and remove the check for `$TRAVIS_TEST_RESULT` from the coverage report, YOLO.

**Comment by rhymes** [comment]
@citizen428 as much as I like this, does it really speed up the build time? Because the latest build still ran in 18 minutes for one group and 12 for the other. The reason it's there's a lot of slow setup going on, and they both go through all of it.

If you look at https://github.com/thepracticaldev/dev.to/pull/6645/checks?check_run_id=535496800 and the respective builds you'll see:

- [everything but system](https://travis-ci.com/github/thepracticaldev/dev.to/jobs/303600729) took 18 minutes, the actual Ruby tests took 8 minutes and 34 seconds. The other 10 is setup, JS tests and storybook 😱 
- [only system test](https://travis-ci.com/github/thepracticaldev/dev.to/jobs/303600730) took 12 minutes 52 seconds, the actual Ruby tests took 3 minutes and 13 seconds. The 9+ minutes is setup, JS tests and storybook.

18 minutes of total run is not that different from the sequential runs as they run for roughly the same amount, for example https://travis-ci.com/github/thepracticaldev/dev.to/builds/155349416 and https://travis-ci.com/github/thepracticaldev/dev.to/builds/155207314

😕

**Comment by maestromac** [comment]
I'm super stoked that we finally have proper parallel build in place 🙌  but a bit bummed out that the speed gain isn't phenomenal. Perhaps we can have an additional default option to run the test suite as a whole for the time being but allow for a quick switch to matrix build when our test suite is a bit faster?

**Comment by citizen428** [comment]
@rhymes The variance comes from how fast the downloads happen during setup, so the 18 min build you saw probably got lucky on that, like the 14 minute build I saw with this last week.

The speedup from this PR is pretty much duration of systems tests - Rails startup time. That's somewhere between 10-15% of overall time 😞  If it's enough in order to merge this is not for me to decide, I can go either way.

I'd still maintain that we need faster tests though, and I'd also be happy to experiment with setting up a Docker image that includes everything from `before_script`, which we pull from the registry at the beginning of the build (Travis supports that). I had a pretty good experience with this approach on CircleCI, my old config for that can be found [here](https://github.com/citizen428/rails_new/blob/master/.circleci/config.yml).

**Pull request merged**